plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/ReusableConnectionFactory.java

```diff
@@ -0,0 +1,184 @@
+import com.google.common.cache.RemovalNotification;
+import io.trino.spi.TrinoException;
+import io.trino.spi.connector.ConnectorSession;
+import org.gaul.modernizer_maven_annotations.SuppressModernizer;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import javax.inject.Inject;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.time.Duration;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.cache.RemovalCause.EXPLICIT;
+import static io.trino.plugin.jdbc.JdbcErrorCode.JDBC_ERROR;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public final class ReusableConnectionFactory
+        implements ConnectionFactory, JdbcQueryEventListener
+{
+    @GuardedBy("this")
+    private final Cache<String, Connection> connections;
+    private final ConnectionFactory delegate;
+
+    @Inject
+    public ReusableConnectionFactory(@ForReusableConnectionFactory ConnectionFactory delegate)
+    {
+        this(delegate, Duration.ofSeconds(2), 10);
+    }
+
+    ReusableConnectionFactory(ConnectionFactory delegate, Duration duration, long maximumSize)
+    {
+        this.connections = createConnectionsCache(duration, maximumSize);
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    // CacheBuilder.build(CacheLoader) is forbidden, because it does not support eviction for ongoing loads.
+    // In this class, loading is not used and cache is used more as a map. So this is safe.
+    @SuppressModernizer
+    private static Cache<String, Connection> createConnectionsCache(Duration duration, long maximumSize)
+    {
+        requireNonNull(duration, "duration is null");
+        return CacheBuilder.newBuilder()
+                .maximumSize(maximumSize)
+                .expireAfterWrite(duration)
+                .removalListener(ReusableConnectionFactory::onRemoval)
+                .build();
+    }
+
+    private static void onRemoval(RemovalNotification<String, Connection> notification)
+    {
+        if (notification.getCause() == EXPLICIT) {
+            // connection was taken from the cache
+            return;
+        }
+        try {
+            requireNonNull(notification.getValue(), "notification.getValue() is null");
+            notification.getValue().close();
+        }
+        catch (SQLException e) {
+            throw new TrinoException(JDBC_ERROR, e);
+        }
+    }
+
+    @Override
+    public Connection openConnection(ConnectorSession session)
+            throws SQLException
+    {
+        String queryId = session.getQueryId();
+        Connection connection = getConnection(session, queryId);
+        return new CachedConnection(queryId, connection);
+    }
+
+    private Connection getConnection(ConnectorSession session, String queryId)
+            throws SQLException
+    {
+        Connection connection = connections.asMap().remove(queryId);
```

Comment:
assumes connections are going to be acquired from a single thread - this is probably true for metadata queries.

Reply:
It is not necessary. This factory can be used by multiple threads. The point is that only one connection can be stored for future use and no connection is not shared between threads.

There are no issues if two threads wants to create a connection for same queryId or single thread that wants to create two connections. This factory is not optimized for such use case so hit rate will be low, but there will be no regression and no correctness issues.